### PR TITLE
Add support for slug transliteration

### PR DESCRIPTION
**Pull request by Zeokat**
TODO: German - GERMANY (de_DE), Danish - DENMARK (da_DK), Serbian - SERBIA (sr_RS)
This languages needs special replacements but will be pending until we have a function to get forum language (locale used with country code maybe).

### CODE REVIEW

File: src/Util/Str.php [hunk 1]
```diff
@@ -14,20 +14,249 @@
...
-     * @param string $str
-     * @return string
+     * @param string $title The string used to generate the slug (usually post title).
+     * @return string The sanitized slug with unicode support.
      */
-    public static function slug($str)
+    public static function slug($title)
     {
+        // Replace accents and create sanitized slug
+        $slug = (new self)->sanitize_slug_string((new self)->convert_accents($title));
```
Comment (Zeokat) [inline src/Util/Str.php]: To use that maybe i have to make all methods of the class static, right? I tryed your code directly and returns an error:

> Deprecated: Non-static method Flarum\Util\Str::sanitize_slug_string() should not be called statically in C:\xampp\htdocs\Str_dev2.php on line 30

I have no idea about how to fix this, i'm not a PHP developer sorry 🤔. If you can modify the code to make it works properly will be good.

File: src/Util/Str.php [hunk 2]
```diff
@@ -14,20 +14,249 @@
...
+    public function sanitize_slug_string($str)
+    {
+        $str = preg_replace('/%([a-fA-F0-9][a-fA-F0-9])/', '----$1----', $str);
+        // Remove % char that are not part of URLencoded chars.
+        $str = str_replace('%', '', $str);
+        // Restore original URLencoded chars with %.
+        $str = preg_replace('/----([a-fA-F0-9][a-fA-F0-9])----/', '%$1', $str);
+
+        if ($this->seems_utf8($str)) {
+            if (function_exists('mb_strtolower')) {
```
Comment (Zeokat) [inline src/Util/Str.php]: Fixed.

File: src/Util/Str.php [hunk 3]
```diff
@@ -14,20 +14,249 @@
...
         $str = strtolower($str);
-        $str = preg_replace('/[^a-z0-9]/i', '-', $str);
+
+        // Simple dot replacement
+        $str = str_replace('.', '-', $str);
+
+        // Keep only normalized chars (including URL encoded chars)
+        $str = preg_replace('/[^%a-z0-9 _-]/', '', $str);
+
+        // Remove nultiple spaces
```
Comment (Zeokat) [inline src/Util/Str.php]: Fixed.

File: src/Util/Str.php [hunk 4]
```diff
@@ -14,20 +14,249 @@
...
+     * Checks if a string is utf8 encoded.
+     * We use strict UTF-8 check with mbstring extension: mb_detect_encoding($str,'UTF-8', true).
+     *
+     *
+     * @param string $str String to be checked against UTF-8 models
+     * @return bool True if string follows a UTF-8 model, false otherwise.
+     */
+    public function seems_utf8($str)
+    {
+        if (mb_detect_encoding($str, 'UTF-8', true)) {
```
Comment (Zeokat) [inline src/Util/Str.php]: Fixed.

File: src/Util/Str.php (hunk 1 above)
Comment (Zeokat) [inline src/Util/Str.php]: I think that finaly everything is coded as expected. Anyway verify please.

**Comment by Zeokat** [comment]
@Luceos files was already changed and fixed? What else? 😕

**Comment by Luceos** [comment]
@Zeokat wait for a response by the merge master @franzliedke

**Comment by Zeokat** [comment]
I go to close it. I need this code, so i just coded an extension to have slug trasliteration. This code will be licensed as GPL as a private extension.

**Comment by tharugossip** [comment]
@Zeokat can you share this extention with flarum community? how to install this extention?

**Pull request closed without merging**